src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala

```diff
@@ -120,23 +135,64 @@ class GoogleDataprocDAO(protected val dataprocConfig: DataprocConfig, protected
 
-    // Create a SoftwareConfig and set a property that makes the cluster have only one node
-    val softwareConfig = new SoftwareConfig().setProperties(Map("dataproc:dataproc.allow.zero.workers" -> "true").asJava)
+    // Create a config for the master node, if properties are not specified in request, use defaults
+    val masterConfig = new InstanceGroupConfig()
+      .setMachineTypeUri(clusterRequest.masterMachineType.getOrElse(clusterDefaultsConfig.masterMachineType))
+      .setDiskConfig(new DiskConfig().setBootDiskSizeGb(clusterRequest.masterDiskSize.getOrElse(clusterDefaultsConfig.masterDiskSize).toInt))
 
     // Create a Cluster Config and give it the GceClusterConfig, the NodeInitializationAction and the SoftwareConfig
     val clusterConfig = new ClusterConfig()
-      .setGceClusterConfig(gce)
-      .setInitializationActions(initActions.asJava).setSoftwareConfig(softwareConfig)
+      .setGceClusterConfig(gceClusterConfig)
+      .setInitializationActions(initActions.asJava)
+      .setMasterConfig(masterConfig)
 
-    // Create a Cluster and give it a name and a Cluster Config
-    val cluster = new DataprocCluster()
-      .setClusterName(clusterName.string)
-      .setConfig(clusterConfig)
+    addWorkerConfigs(clusterConfig, clusterRequest)
+  }
 
-    // Create a dataproc create request and give it the google project, a zone, and the Cluster
-    val request = dataproc.projects().regions().clusters().create(googleProject.string, dataprocConfig.dataprocDefaultRegion, cluster)
+  private def addWorkerConfigs(clusterConfig: ClusterConfig, clusterRequest: ClusterRequest): ClusterConfig = {
+    // If the number of workers is zero, make a Single Node cluster, else make a Standard one
+    if (clusterRequest.numberOfWorkers.getOrElse(clusterDefaultsConfig.numberOfWorkers) == 0) {
+      // Create a SoftwareConfig and set a property that makes the cluster have only one node
+      val softwareConfig = new SoftwareConfig().setProperties(Map("dataproc:dataproc.allow.zero.workers" -> "true").asJava)
+      clusterConfig.setSoftwareConfig(softwareConfig)
+    }
+    else // Standard, multi node cluster
+      getStandardConfig(clusterConfig, clusterRequest)
+  }
 
-    executeGoogleRequestAsync(googleProject, clusterName.toString, request)  // returns a Future[DataprocOperation]
+  private def getStandardConfig(clusterConfig: ClusterConfig, clusterRequest: ClusterRequest): ClusterConfig = {
+    // Set the configs of the non-preemptible, primary worker nodes
+    val clusterConfigWithWorkerConfigs = clusterConfig.setWorkerConfig(getPrimaryWorkerConfig(clusterRequest))
+
+    // If the number of preemptible workers is greater than 0, set a secondary worker config
+    val numberOfPreemptibleWorkers = clusterRequest.numberOfPreemptibleWorkers.getOrElse(clusterDefaultsConfig.numberOfPreemptibleWorkers)
```

Comment:
This is a bit pie in the sky but I wonder if there's a different way to do this than lots of `getOrElse`. Essentially what you want to do is take the default config and apply the user's in such a way that a `None` value in the user config doesn't overwrite the default config but does everywhere else.

Some googling says that Shapeless can do this, but the code I've found it kinda complex. I know the config loader system we have handles composing multiple config files on top of each other.

@rtitle any thoughts?

Reply:
I'm sure it's possible with Shapeless (`Poly`?) -- I'm not the biggest shapeless expert though. 

Another idea I had would be to use a semigroup. This would require you use the same case class for your real values and default values.
```
scala> case class Config(a: Option[Int], b: Option[String])
defined class Config

scala> implicit val configSemigroup = new Semigroup[Config] {
     | def combine(config1: Config, config2: Config): Config = 
     | Config(config1.a.orElse(config2.a), config1.b.orElse(config2.b))
     | }
configSemigroup: cats.Semigroup[Config] = $anon$1@4dffb0e9

scala> val defaults = Config(Some(3), Some("foo"))
defaults: Config = Config(Some(3),Some(foo))

scala> val real = Config(Some(4), None)
real: Config = Config(Some(4),None)

scala> real |+| defaults
res0: Config = Config(Some(4),Some(foo))
```